### ISSUE TITLE
[java] Suspicious false positives in AssignmentInOperand Rule

### ISSUE THREAD

**Issue by Noah0120** [pmd/pmd#3429]
**Affects PMD Version:** 
6.36.0
**Rule:**
AssignmentInOperand
Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/latest/pmd_rules_java_errorprone.html#assignmentinoperand

**Description:**
I got suspicious false positives from AssignmentInOperand rule 
These detection results weren't `assignment` in an `operand`.

There are five cases:
1) `assignment` in a `method argument`
2) `assignment` in a `conditional expression`
3) `assignment` in an `array index`
4) `assignment` in a `method caller`
5) `assignment` in a `constructor`

For these cases, I wondered 
you assumed the above five cases as an `Operand`.
Also, I thought that it is better to make a new rule for each case.
What do you think?

**Code Sample demonstrating the issue:**

```
class AssignmentInOperand{
        void test(){
        while (keys.contains(k = random.nextLong())); //case1
        for (j = i; Character.isJavaIdentifierPart(text.charAt(--j));); //case1

        if ( addNextStaticMember = (staticMember.getNextStaticMember() != null) ? true : false ); //case2
        for (Method m : (null == theClassMethods ? theClassMethods = theClass.getMethods() : theClassMethods)); //case2

        while (Character.isWhitespace(segment.array[offset++])); //case3
        if (subPropMatches[i++]); //case3

        while ((is = ia.getNext(null)).isPresent()); //case4
        if ((destFile = new File (dest, srcFO.getNameExt ())).exists ()); //case4

        while (new File(location, projectName = projectNamePrefix + "_" + num).exists()); //case5
        }
}
```

**Expected outcome:**
All of the case 1 to case 5 are suspicious false positives
because the assignments are not in operand
but in `method argument`, 
`conditional expression`(just assigning from conditional exp or then branch),
`array index`, `method caller`, and `constructor`.

Thank you for your time and consideration.

**Running PMD through:** *[CLI]*

**Comment by oowekyala** [comment]
Hi, I think there's a misunderstanding about the meaning of "operand". An "operand" here is a subexpression of another expression. This means that the result of the expression will immediately be used by another expression, as opposed to when the assignment is in statement context, where the expression result is ignored. All the cases you listed fit this definition, so this is not a false positive.

**Comment by Noah0120** [comment]
> An "operand" here is a subexpression of another expression. 

I'm confused about the meaning of `operand` here.

> This means that the result of the expression will immediately be used by another expression, 

Does `the result of the expression` mean the result of the assignment?
For example,
In `while (keys.contains(k = random.nextLong())); //case1`,
`k=random.nextLong()` will immediately be used by `keys.contains()`?
Or `k=random.nextLong()` will immediately be used by `conditional expression` of `while stmt`?
Then, does the rule try to detect `assignment` in `conditional expression`, not the things that used with operators? 

> as opposed to when the assignment is in statement context, where the expression result is ignored. 

Could you give me an example of the `ignored case(when the assignment is in statement context)`?

**Comment by oowekyala** [comment]
> In `while (keys.contains(k = random.nextLong()));`,  `k=random.nextLong() ` will immediately be used by `keys.contains()`?

Yes, exactly this. 

Java allows assignments to be written everywhere an expression is expected, including in eg method arguments, like in `keys.contains(k = ...)`. The rule enforces a code style where assignments are only allowed in statement position, eg `k = ...;`.

For instance
```java
 for (j = i; Character.isJavaIdentifierPart(text.charAt(--j));); //case1
```
can be rewritten to 
```java
 for (j = i - 1; Character.isJavaIdentifierPart(text.charAt(j)); j--); //case1
```
The update clause of a for loop is also a "statement position". Nothing uses the value of the assignment *expression* itself, only the value of the *variable*. 

Another example:
```java
System.out.println(array[++i]);
```
can be rewritten to 
```java
i++;
System.out.println(array[i]);
```

**Comment by Noah0120** [comment]
Oh, I got it! Thank you for your kind explanation:)

Then, I have one more question.
Now, PMD doesn't detect the statements such as 
`key.contains(k=random.nextLong());` and `Character.isJavaIdentifierPart(text.charAt(--j));`
when they are located out side of conditional expression of for-stmt, while-stmt, if-stmt.
It means:
````
void test(){
        key.contains(k=random.nextLong());
        Character.isJavaIdentifierPart(text.charAt(--j));
}
````
the code above doesn't generate any alarms.
As you said, it should be fixed as
````
void test(){
        k=random.nextLong();
        key.contains(k);
        --j;
        Character.isJavaIdentifierPart(text.charAt(j));
}
````

Can we regard them as false negatives?

**Comment by oowekyala** [comment]
Yes I would say they are. Can you please open a new issue for those?

**Comment by Noah0120** [comment]
Sure:)
